src/fees/compute_fees.py

```diff
@@ -0,0 +1,599 @@
+                fee_policies.append(
+                    VolumeFeePolicy(Fraction(fee_policy["volume"]["factor"]))
+                )
+            elif "priceImprovement" in fee_policy:
+                quote = Quote(
+                    int(fee_policy["priceImprovement"]["quote"]["sellAmount"]),
+                    int(fee_policy["priceImprovement"]["quote"]["buyAmount"]),
+                    int(fee_policy["priceImprovement"]["quote"]["fee"]),
+                )
+                fee_policies.append(
+                    PriceImprovementFeePolicy(
+                        Fraction(fee_policy["priceImprovement"]["factor"]),
+                        Fraction(fee_policy["priceImprovement"]["maxVolumeFactor"]),
+                        quote,
+                    )
+                )
+            else:
+                raise ValueError(f"Fee kind {fee_policy.keys()} is invalid.")
+        return fee_policies
+
+
+def fetch_settlement_data(
+    tx_hash: HexBytes,
+) -> tuple[OnchainSettlementData, OffchainSettlementData]:
+    onchain_fetcher = BlockchainFetcher()
+    offchain_fetcher = OrderbookAWSFetcher()
+
+    onchain_data = onchain_fetcher.get_onchain_data(tx_hash)
+    offchain_data = offchain_fetcher.get_offchain_data(onchain_data)
+
+    return onchain_data, offchain_data
+
+
+# computing fees
+
+
+def compute_fee_imbalances(
+    onchain_data: OnchainSettlementData, offchain_data: OffchainSettlementData
+) -> tuple[dict[HexBytes, int], dict[HexBytes, int]]:
+    protocol_fees: dict[HexBytes, int] = {}
+    network_fees: dict[HexBytes, int] = {}
+    for trade in onchain_data.trades:
+        # protocol fees
+        fee_policies = offchain_data.trade_fee_policies[trade.order_uid]
+        protocol_fee_amount = trade.protocol_fee(fee_policies)
+        protocol_fee_token = trade.surplus_token()
+        protocol_fees[protocol_fee_token] = protocol_fee_amount
+
+        # network fees
+        surplus_fee = trade.compute_surplus_fee()  # in the surplus token
+        network_fee = surplus_fee - protocol_fee_amount
+        if trade.kind == "sell":
+            network_fee_sell = int(
+                network_fee
+                * Fraction(
+                    trade.buy_token_clearing_price, trade.sell_token_clearing_price
+                )
+            )
+        else:
+            network_fee_sell = network_fee
+
+        network_fees[trade.sell_token] = network_fee_sell
+
+    return protocol_fees, network_fees
+
+
+# combined function
+
+
+def batch_fee_imbalances(
+    tx_hash: HexBytes,
+) -> tuple[dict[str, int], dict[str, int]]:
+    onchain_data, offchain_data = fetch_settlement_data(tx_hash)
+    protocol_fees, network_fees = compute_fee_imbalances(onchain_data, offchain_data)
+    protocol_fees = {
+        token.hex(): fee
+        for token, fee in protocol_fees.items()
+    }
+    network_fees = {
+        token.hex(): fee
+        for token, fee in network_fees.items()
+    }
+    return protocol_fees, network_fees
+
```

Comment:
To fix the mypy issue, I would change the return type of `compute_fee_imbalances` to use `str` and remove the conversion from `batch_fee_imbalances`. The error was due to variables changing types in `batch_fee_imbalances`.

```suggestion
def compute_fee_imbalances(
    onchain_data: OnchainSettlementData, offchain_data: OffchainSettlementData
) -> tuple[dict[str, int], dict[str, int]]:
    protocol_fees: dict[str, int] = {}
    network_fees: dict[str, int] = {}
    for trade in onchain_data.trades:
        # protocol fees
        fee_policies = offchain_data.trade_fee_policies[trade.order_uid]
        protocol_fee_amount = trade.protocol_fee(fee_policies)
        protocol_fee_token = trade.surplus_token()
        protocol_fees[protocol_fee_token.hex()] = protocol_fee_amount

        # network fees
        surplus_fee = trade.compute_surplus_fee()  # in the surplus token
        network_fee = surplus_fee - protocol_fee_amount
        if trade.kind == "sell":
            network_fee_sell = int(
                network_fee
                * Fraction(
                    trade.buy_token_clearing_price, trade.sell_token_clearing_price
                )
            )
        else:
            network_fee_sell = network_fee

        network_fees[trade.sell_token.hex()] = network_fee_sell

    return protocol_fees, network_fees


# combined function


def batch_fee_imbalances(
    tx_hash: HexBytes,
) -> tuple[dict[str, int], dict[str, int]]:
    onchain_data, offchain_data = fetch_settlement_data(tx_hash)
    protocol_fees, network_fees = compute_fee_imbalances(onchain_data, offchain_data)
    return protocol_fees, network_fees

```